src/engine/execute.js

```diff
@@ -173,37 +233,51 @@ const execute = function (sequencer, thread) {
     // Generate values for arguments (inputs).
     const argValues = {};
 
-    // Add all fields on this block to the argValues.
-    for (const fieldName in fields) {
-        if (!fields.hasOwnProperty(fieldName)) continue;
-        if (fieldName === 'VARIABLE' || fieldName === 'LIST' ||
-            fieldName === 'BROADCAST_OPTION') {
-            argValues[fieldName] = {
-                id: fields[fieldName].id,
-                name: fields[fieldName].value
-            };
-        } else {
+    // Add all fields on this block to the argValues. Some known fields may
+    // appear by themselves and can be set to argValues quicker by setting them
+    // explicitly.
+    if (blockCached._isFieldKnown) {
```

Comment:
It looks like, with minor modification above, this could be simplified to something like:
```js
switch (blockCached._fieldKind) {
case foo.FIELD_VARIABLE:
    argValues.VARIABLE = blockCached._fieldVariable;
    break;
// similar cases for list & broadcast
default:
    for (const fieldName in fields) {
...
```